src/core/parse.js

```diff
@@ -635,19 +725,69 @@ foam.CLASS({
         s: this.s,
         success: success,
         fail: fail
-      })
+      });
     },
 
-    function parse(ps, obj) {
+    function parse(ps) {
       var str = this.s;
       for ( var i = 0 ; i < str.length ; i++, ps = ps.tail ) {
-        if ( str.charAt(i) !== ps.head ) return undefined;
+        if ( ! ps.head || str.charAt(i) !== ps.head ) {
```

Comment:
I don't see why all these ! ps.head checks were added to the literal parsers.  FOAM1 parsers did not include them.